libsmtutil/SMTLib2Context.cpp

```diff
@@ -0,0 +1,316 @@
+	{
+		size_t size = std::stoul(_expr.arguments[1].name);
+		auto arg = toSExpr(_expr.arguments.front());
+		auto int2bv = "(_ int2bv " + std::to_string(size) + ")";
+		// Some solvers treat all BVs as unsigned, so we need to manually apply 2's complement if needed.
+		sexpr += std::string("ite ") +
+			"(>= " + arg + " 0) " +
+			"(" + int2bv + " " + arg + ") " +
+			"(bvneg (" + int2bv + " (- " + arg + ")))";
+	}
+	else if (_expr.name == "bv2int")
+	{
+		auto intSort = std::dynamic_pointer_cast<IntSort>(_expr.sort);
+		smtAssert(intSort, "");
+
+		auto arg = toSExpr(_expr.arguments.front());
+		auto nat = "(bv2nat " + arg + ")";
+
+		if (!intSort->isSigned)
+			return nat;
+
+		auto bvSort = std::dynamic_pointer_cast<BitVectorSort>(_expr.arguments.front().sort);
+		smtAssert(bvSort, "");
+		auto size = std::to_string(bvSort->size);
+		auto pos = std::to_string(bvSort->size - 1);
+
+		// Some solvers treat all BVs as unsigned, so we need to manually apply 2's complement if needed.
+		sexpr += std::string("ite ") +
+			"(= ((_ extract " + pos + " " + pos + ")" + arg + ") #b0) " +
+			nat + " " +
+			"(- (bv2nat (bvneg " + arg + ")))";
+	}
+	else if (_expr.name == "const_array")
+	{
+		smtAssert(_expr.arguments.size() == 2, "");
+		auto sortSort = std::dynamic_pointer_cast<SortSort>(_expr.arguments.at(0).sort);
+		smtAssert(sortSort, "");
+		auto arraySort = std::dynamic_pointer_cast<ArraySort>(sortSort->inner);
+		smtAssert(arraySort, "");
+		sexpr += "(as const " + toSmtLibSort(arraySort) + ") ";
+		sexpr += toSExpr(_expr.arguments.at(1));
+	}
+	else if (_expr.name == "tuple_get")
+	{
+		smtAssert(_expr.arguments.size() == 2, "");
+		auto tupleSort = std::dynamic_pointer_cast<TupleSort>(_expr.arguments.at(0).sort);
+		size_t index = std::stoul(_expr.arguments.at(1).name);
+		smtAssert(index < tupleSort->members.size(), "");
+		sexpr += "|" + tupleSort->members.at(index) + "| " + toSExpr(_expr.arguments.at(0));
+	}
+	else if (_expr.name == "tuple_constructor")
+	{
+		auto tupleSort = std::dynamic_pointer_cast<TupleSort>(_expr.sort);
+		smtAssert(tupleSort, "");
+		sexpr += "|" + tupleSort->name + "|";
+		for (auto const& arg: _expr.arguments)
+			sexpr += " " + toSExpr(arg);
+	}
+	else
+	{
+		sexpr += _expr.name;
+		for (auto const& arg: _expr.arguments)
+			sexpr += " " + toSExpr(arg);
+	}
+	sexpr += ")";
+	return sexpr;
+}
+
+std::optional<SortPointer> SMTLib2Context::getTupleType(std::string const& _name) const
+{
+	auto it = m_tupleSorts.find(_name);
+	return it == m_tupleSorts.end() ? std::nullopt : std::optional<SortPointer>(unresolve(it->second));
+}
+
+std::optional<std::pair<std::string, SortPointer>> SMTLib2Context::getTupleAccessor(std::string const& _name) const
+{
+	for (auto&& [_, sortId] : m_tupleSorts)
+	{
+		auto const& type = m_knownTypes.at(sortId.id);
+		smtAssert(type->kind == Kind::Tuple);
+		auto const& tupleType = dynamic_cast<TupleType const&>(*type);
+		for (auto&& [memberName, memberSort] : tupleType.m_accessors)
+		{
```

Comment:
There is only one expression in `for` body. We can remove the curly braces.

Reply:
Fixed.